Formal2024/Workshops/workshop3.lean

```diff
@@ -0,0 +1,210 @@
+  rcases h with ⟨h₀, h₁⟩
+  contrapose! h₁
+  exact le_antisymm h₀ h₁
+
+example {x y : ℝ} : x ≤ y ∧ x ≠ y → ¬y ≤ x := by
+  rintro ⟨h₀, h₁⟩ h'
+  exact h₁ (le_antisymm h₀ h')
+
+-- You can also access the parts of a conjunction `h` with `h.left` and `h.right` (or `h.1` and `h.2`).
+example {x y : ℝ} (h : x ≤ y ∧ x ≠ y) : ¬y ≤ x :=
+  fun h' ↦ h.right (le_antisymm h.left h')
+
+-- You can next anonymous constructors, even if they mix conjunctions and existential quantifiers.
+-- Note that they (usually) associate to the right.
+example : ∃ x : ℕ, 3 < x ∧ x < 5 ∧ ∃ y, y^2 = x :=
+  ⟨4, by norm_num, by norm_num, 2, by norm_num⟩
+
+
+/-!
+  # Iff
+-/
+
+-- The symbol for iff is `↔`, which you can write as `\iff`.
+
+-- As with `∧`, you can use the `constructor` tactic, or the anonymous constructor `⟨⟩` to prove an iff.
+-- Similarly, you can use an iff with `rcases` or `rintro`.
+-- If `h` is an iff, you can access the two implications with `h.mp` and `h.mpr` (or `h.1` and `h.2`).
+
+example {x y : ℝ} (h : x ≤ y) : ¬y ≤ x ↔ x ≠ y := by
+  constructor
+  · intro h' heq
+    apply h'
+    rw [heq]
+  · intro h' hle
+    apply h'
+    exact le_antisymm h hle
+
+
+-- Lean knows that `↔` is reflexive, symmetric, and transitive, so you can use the `rw` tactic with it.
+#check abs_lt
+
+example (x : ℝ) : |x + 3| < 5 → -8 < x ∧ x < 2 := by
+  rw [abs_lt]
+  intro h
+  constructor <;> linarith
+-- The `<;>` tactic combinator attempts to close every open goal with the given tactic
+
+
+/-!
+  # Exercises
+-/
+
+-- 1. Try to prove this lemma in at least two different ways, e.g. using `constructor`, `⟨⟩`, `h.left` and `h.right`, etc. [MIL 3.4]
+
+example {x y : ℝ} : x ≤ y ∧ ¬y ≤ x ↔ x ≤ y ∧ x ≠ y := by
+  sorry
+
+
+-- 2. Prove the following lemma. [MIL 3.4]
+
+example : ¬Monotone fun x : ℝ ↦ -x := by
+  rw [Monotone] -- This unfolds the definition of `Monotone` in the goal.
+  push_neg
+  sorry
+
+
+/-!
+  # Disjunction
+-/
+
+-- The symbol for disjunction is `∨`, which you can write as `\or`.
+
+-- To prove a disjunction, it is enough to prove one of the two sides.
+-- You can use the `left` or `right` to choose which side you want to prove.
+
+example {x y : ℝ} (h : y > x ^ 2) : y > 0 ∨ y < -1 := by
+  left
+  linarith [pow_two_nonneg x] -- You can add extra hypotheses to `linarith` with square brackets.
+  -- In this case we have added `x^2 ≥ 0`.
+
+-- The proof term version of this is `Or.inl` and `Or.inr`.
+example {x y : ℝ} (h : y > x ^ 2) : y > 0 ∨ y < -1 :=
+  Or.inl (by linarith [pow_two_nonneg x])
+
+
+-- To use a disjunction, you can use `rcases` or `rintro`, as with existential quantifiers and conjunctions.
+-- The notation is slightly different: you use `|` instead of `⟨⟩`.
+-- When we use a disjunction, we have to prove the goal for each side of the disjunction.
+example {x y : ℝ} : x < |y| → x < y ∨ x < -y := by
+  rcases le_or_gt 0 y with h | h -- `le_or_gt` is a lemma that says `0 ≤ y ∨ y < 0`.
```

Comment:
"Notice that we can call the hypotheses introduced by rcases with the same name, as they will be used in two separate cases"